README.md
# SwiftLint

An experimental tool to enforce Swift style and conventions, loosely based on
[GitHub's Swift Style Guide](https://github.com/github/swift-style-guide).

SwiftLint hooks into [Clang](http://clang.llvm.org) and
[SourceKit](http://www.jpsim.com/uncovering-sourcekit) to use the
[AST](http://clang.llvm.org/docs/IntroductionToTheClangAST.html) representation
of your source files for more accurate results.

![Test Status](https://travis-ci.org/realm/SwiftLint.svg?branch=master)

## Installation

Using [Homebrew](http://brew.sh/)

```
brew install swiftlint
```

You can also install SwiftLint by downloading `SwiftLint.pkg` from the
[latest GitHub release](https://github.com/realm/SwiftLint/releases/latest) and
running it.

You can also build from source by cloning this project and running
`make install`.

## Usage

### Xcode

Integrate SwiftLint into an Xcode scheme to get warnings and errors displayed
in the IDE. Just add a new "Run Script Phase" with `/usr/local/bin/swiftlint`
as its contents.

![](screenshot.png)

### Atom

To integrate SwiftLint with [Atom](https://atom.io/) install the
[`linter-swiftlint`](https://atom.io/packages/linter-swiftlint) package from
APM.

### Command Line

```
$ swiftlint help
Available commands:

   help      Display general or command-specific help
   lint      Print lint warnings and errors for the Swift files in the current directory (default command)
   rules     Display the list of rules and examples
   version   Display the current version of SwiftLint
```

Run `swiftlint` in the directory containing the Swift files to lint. Directories
will be searched recursively.

## Rules

There are only a small number of rules currently implemented, but we hope the
Swift community (that's you!) will contribute more over time. Pull requests are
encouraged.

The rules that *are* currently implemented are mostly there as a starting point
and are subject to change.

See the [Source/SwiftLintFramework/Rules](Source/SwiftLintFramework/Rules)
directory to see the currently implemented rules.

### Configuration

Configure SwiftLint by adding a `.swiftlint.yml` file from the directory you'll
run SwiftLint from. The following parameters can be configured:

```yaml
disabled_rules: # rule identifiers to exclude from running
  - todo
included: # paths to include during linting. `--path` is ignored if present. takes precendence over `excluded`.
  - Source
excluded: # paths to ignore during linting. overridden by `included`.
  - Carthage
  - Pods
```

## License

MIT licensed.


CHANGELOG.md
## Master

##### Breaking

* None.

##### Enhancements

* Added `OperatorFunctionWhitespaceRule` to make sure that
  you use whitespace around operators when defining them.  
  [Akira Hirakawa](https://github.com/akirahrkw)
  [#60](https://github.com/realm/SwiftLint/issues/60)

* Added `ReturnArrowWhitespaceRule` to make sure that
  you have 1 space before return arrow and return type.  
  [Akira Hirakawa](https://github.com/akirahrkw)

* Support linting from standard input (use `--use-stdin`).  
  [JP Simard](https://github.com/jpsim)
  [#78](https://github.com/realm/SwiftLint/issues/78)

* Improve performance of `TrailingNewlineRule`.  
  [Keith Smiley](https://github.com/keith)

* Configure SwiftLint via a YAML file:
  Supports `disabled_rules`, `included` and `excluded`.
  Pass a configuration file path to `--config`, defaults to `.swiftlint.yml`.  
  [JP Simard](https://github.com/jpsim)
  [#3](https://github.com/realm/SwiftLint/issues/3)

##### Bug Fixes

* None.


## 0.1.1: Top Loading

##### Breaking

* The `Rule` and `ASTRule` protocol members are now non-static.  
  [aarondaub](https://github.com/aarondaub)

* Split `Rule` into `Rule` and `ParameterizedRule` protocols.  
  [aarondaub](https://github.com/aarondaub)
  [#21](https://github.com/realm/SwiftLint/issues/21)

##### Enhancements

* Added a command line option `--path` to specify a path to lint.  
  [Lars Lockefeer](https://github.com/larslockefeer)
  [#16](https://github.com/realm/SwiftLint/issues/16)

* swiftlint now returns a non-zero error code when a warning of high-severity
  or above is found in the source files being linted.  
  [Pat Wallace](https://github.com/pawrsccouk)
  [#30](https://github.com/realm/SwiftLint/issues/30)

* Added `rules` command to display which rules are currently applied along
  with examples.  
  [Chris Eidhof](https://github.com/chriseidhof)

* Cache parsing to reduce execution time by more than 50%.  
  [Nikolaj Schumacher](https://github.com/nschum)

* Added `ControlStatementRule` to make sure that if/for/while/do statements
  do not wrap their conditionals in parentheses.  
  [Andrea Mazzini](https://github.com/andreamazz)

* Character position is now included in violation location where appropriate.  
  [JP Simard](https://github.com/jpsim)
  [#62](https://github.com/realm/SwiftLint/issues/62)

* The following rules now conform to `ASTRule`:
  FunctionBodyLength, Nesting, TypeBodyLength, TypeName, VariableName.  
  [JP Simard](https://github.com/jpsim)

##### Bug Fixes

* Trailing newline and file length violations are now displayed in Xcode.  
  [JP Simard](https://github.com/jpsim)
  [#43](https://github.com/realm/SwiftLint/issues/43)


## 0.1.0: Fresh Out Of The Dryer

First Version!


Source/SwiftLintFramework/Rules/NestingRule.swift
//
//  NestingRule.swift
//  SwiftLint
//
//  Created by JP Simard on 2015-05-16.
//  Copyright (c) 2015 Realm. All rights reserved.
//

import SourceKittenFramework
import SwiftXPC

public struct NestingRule: ASTRule {
    public init() {}

    public let identifier = "nesting"

    public func validateFile(file: File) -> [StyleViolation] {
        return validateFile(file, dictionary: file.structure.dictionary)
    }

    public func validateFile(file: File, dictionary: XPCDictionary) -> [StyleViolation] {
        let substructure = dictionary["key.substructure"] as? XPCArray ?? []
        return substructure.flatMap { subItem -> [StyleViolation] in
            var violations = [StyleViolation]()
            if let subDict = subItem as? XPCDictionary,
                let kindString = subDict["key.kind"] as? String,
                let kind = SwiftDeclarationKind(rawValue: kindString) {
                    violations.extend(self.validateFile(file, dictionary: subDict))
                    violations.extend(self.validateFile(file, kind: kind, dictionary: subDict))
            }
            return violations
        }
    }

    public func validateFile(file: File,
        kind: SwiftDeclarationKind,
        dictionary: XPCDictionary) -> [StyleViolation] {
        return validateFile(file, kind: kind, dictionary: dictionary, level: 0)
    }

    func validateFile(file: File,
        kind: SwiftDeclarationKind,
        dictionary: XPCDictionary,
        level: Int) -> [StyleViolation] {
        var violations = [StyleViolation]()
        let typeKinds: [SwiftDeclarationKind] = [
            .Class,
            .Struct,
            .Typealias,
            .Enum,
            .Enumcase
        ]
        if let offset = (dictionary["key.offset"] as? Int64).flatMap({ Int($0) }) {
            if level > 1 && typeKinds.contains(kind) {
                violations.append(StyleViolation(type: .Nesting,
                    location: Location(file: file, offset: offset),
                    reason: "Types should be nested at most 1 level deep"))
            } else if level > 2 && kind == .Enumelement {
                // Enum elements are implicitly wrapped in an .Enumcase
                violations.append(StyleViolation(type: .Nesting,
                    location: Location(file: file, offset: offset),
                    reason: "Types should be nested at most 1 level deep"))
            } else if level > 5 {
                violations.append(StyleViolation(type: .Nesting,
                    location: Location(file: file, offset: offset),
                    reason: "Statements should be nested at most 5 levels deep"))
            }
        }
        let substructure = dictionary["key.substructure"] as? XPCArray ?? []
        violations.extend(substructure.flatMap { subItem in
            let subDict = subItem as? XPCDictionary
            let kindString = subDict?["key.kind"] as? String
            let kind = kindString.flatMap { kindString in
                return SwiftDeclarationKind(rawValue: kindString)
            }
            if let kind = kind, subDict = subDict {
                return (kind, subDict)
            }
            return nil
        }.flatMap { (kind, dict) -> [StyleViolation] in
            return self.validateFile(file, kind: kind, dictionary: dict, level: level + 1)
        })
        return violations
    }

    public let example = RuleExample(
        ruleName: "Nesting Rule",
        ruleDescription: "Types should be nested at most 1 level deep, " +
        "and statements should be nested at most 5 levels deep.",
        nonTriggeringExamples: ["class", "struct", "enum"].flatMap { kind in
            ["\(kind) Class0 { \(kind) Class1 {} }\n",
                "func func0() {\nfunc func1() {\nfunc func2() {\nfunc func3() {\nfunc func4() { " +
                "func func5() {\n}\n}\n}\n}\n}\n}\n"]
        },
        triggeringExamples: ["class", "struct", "enum"].map { kind in
            "\(kind) Class0 { \(kind) Class1 { \(kind) Class2 {} } }\n"
            } + [
            "func func0() {\nfunc func1() {\nfunc func2() {\nfunc func3() {\nfunc func4() { " +
            "func func5() {\nfunc func6() {\n}\n}\n}\n}\n}\n}\n}\n"
            ]
    )
}


Source/SwiftLintFramework/Rules/ControlStatementRule.swift
//
//  ControlStatementRule.swift
//  SwiftLint
//
//  Created by Andrea Mazzini on 26/05/15.
//  Copyright (c) 2015 Realm. All rights reserved.
//

import SourceKittenFramework

public struct ControlStatementRule: Rule {
    public init() {}

    public let identifier = "control_statement"

    public func validateFile(file: File) -> [StyleViolation] {
        return ["if", "for", "while"].flatMap { statementKind -> [StyleViolation] in
            let pattern = "\(statementKind)\\s*\\([^,]*\\)\\s*\\{"
            return file.matchPattern(pattern).flatMap { match, syntaxKinds in
                if syntaxKinds.first != .Keyword {
                    return nil
                }
                return StyleViolation(type: .ControlStatement,
                    location: Location(file: file, offset: match.location),
                    severity: .Low,
                    reason: "\(statementKind) statements shouldn't wrap their conditionals in " +
                    "parentheses.")
                }
        }
    }

    public let example = RuleExample(
        ruleName: "Control Statement",
        ruleDescription: "if,for,while,do statements shouldn't wrap their conditionals in " +
        "parentheses.",
        nonTriggeringExamples: [
            "if condition {\n",
            "if (a, b) == (0, 1) {\n",
            "if renderGif(data) {\n",
            "renderGif(data)\n",
            "for item in collection {\n",
            "for (key, value) in dictionary {\n",
            "for (index, value) in enumerate(array) {\n",
            "for var index = 0; index < 42; index++ {\n",
            "while condition {\n",
            "} while condition {\n",
            "do { ; } while condition {\n"
        ],
        triggeringExamples: [
            "if (condition) {\n",
            "if(condition) {\n",
            "for (item in collection) {\n",
            "for (var index = 0; index < 42; index++) {\n",
            "for(item in collection) {\n",
            "for(var index = 0; index < 42; index++) {\n",
            "while (condition) {\n",
            "while(condition) {\n",
            "} while (condition) {\n",
            "} while(condition) {\n",
            "do { ; } while(condition) {\n",
            "do { ; } while (condition) {\n"
        ]
    )
}


Source/SwiftLintFramework/Rules/OperatorFunctionWhitespaceRule.swift
//
//  OperatorWhitespaceRule.swift
//  SwiftLint
//
//  Created by Akira Hirakawa on 8/6/15.
//  Copyright (c) 2015 Realm. All rights reserved.
//

import SourceKittenFramework

public struct OperatorFunctionWhitespaceRule: Rule {
    public init() {}

    public let identifier = "operator_whitespace"

    public func validateFile(file: File) -> [StyleViolation] {
        let operators = ["/", "=", "-", "+", "!", "*", "|", "^", "~", "?", "."].map({"\\\($0)"}) +
            ["%", "<", ">", "&"]
        let zeroOrManySpaces = "(\\s{0}|\\s{2,})"
        let pattern1 = "func\\s+[" + "".join(operators) + "]+\(zeroOrManySpaces)(<[A-Z]+>)?\\("
        let pattern2 = "func\(zeroOrManySpaces)[" + "".join(operators) + "]+\\s+(<[A-Z]+>)?\\("
        return file.matchPattern("(\(pattern1)|\(pattern2))").filter { _, syntaxKinds in
            return syntaxKinds.first == .Keyword
        }.map { range, _ in
            return StyleViolation(type: .OperatorFunctionWhitespace,
                location: Location(file: file, offset: range.location),
                severity: .Medium,
                reason: example.ruleDescription)
        }
    }

    public let example = RuleExample(
        ruleName: "Operator Function Whitespace Rule",
        ruleDescription: "Use a single whitespace around operators when " +
            "defining them.",
        nonTriggeringExamples: [
            "func <| (lhs: Int, rhs: Int) -> Int {}\n",
            "func <|< <A>(lhs: A, rhs: A) -> A {}\n",
            "func abc(lhs: Int, rhs: Int) -> Int {}\n"
        ],
        triggeringExamples: [
            "func <|(lhs: Int, rhs: Int) -> Int {}\n",   // no spaces after
            "func <|<<A>(lhs: A, rhs: A) -> A {}\n",     // no spaces after
            "func <|  (lhs: Int, rhs: Int) -> Int {}\n", // 2 spaces after
            "func <|<  <A>(lhs: A, rhs: A) -> A {}\n",   // 2 spaces after
            "func  <| (lhs: Int, rhs: Int) -> Int {}\n", // 2 spaces before
            "func  <|< <A>(lhs: A, rhs: A) -> A {}\n"    // 2 spaces before
        ]
    )
}


Source/swiftlint/LintCommand.swift
//
//  LintCommand.swift
//  SwiftLint
//
//  Created by JP Simard on 2015-05-16.
//  Copyright (c) 2015 Realm. All rights reserved.
//

import Commandant
import Foundation
import Result
import SourceKittenFramework
import SwiftLintFramework

let fileManager = NSFileManager.defaultManager()

struct LintCommand: CommandType {
    let verb = "lint"
    let function = "Print lint warnings and errors for the Swift files in the current directory " +
                   "(default command)"

    func run(mode: CommandMode) -> Result<(), CommandantError<()>> {
        return LintOptions.evaluate(mode).flatMap { options in
            let configuration = Configuration(path: options.configurationFile,
                optional: !Process.arguments.contains("--config"))
            if options.useSTDIN {
                let standardInput = NSFileHandle.fileHandleWithStandardInput()
                let stdinData = standardInput.readDataToEndOfFile()
                let stdinNSString = NSString(data: stdinData, encoding: NSUTF8StringEncoding)
                if let stdinString = stdinNSString as? String {
                    let file = File(contents: stdinString)
                    let linter = Linter(file: file, configuration: configuration)
                    print("\n".join(linter.styleViolations.map { $0.description }))
                    return .Success()
                }
                return .Failure(CommandantError<()>.CommandError())
            }

            // Otherwise parse path.
            return lint(options.path, configuration: configuration)
        }
    }

    private func lint(path: String,
        configuration: Configuration) -> Result<(), CommandantError<()>> {
        let filesToLint = (configuration.included.count == 0 ? filesToLintAtPath(path) : [])
            .filter({ !configuration.excluded.flatMap(filesToLintAtPath).contains($0) }) +
            configuration.included.flatMap(filesToLintAtPath)
        if filesToLint.count > 0 {
            if path.isEmpty {
                print("Linting Swift files in current working directory")
            } else {
                print("Linting Swift files at path \(path)")
            }
            var numberOfViolations = 0, numberOfSeriousViolations = 0
            for (index, path) in filesToLint.enumerate() {
                let filename = (path as NSString).lastPathComponent
                print("Linting '\(filename)' (\(index + 1)/\(filesToLint.count))")
                let file = File(path: path)!
                for violation in Linter(file: file, configuration: configuration).styleViolations {
                    print(violation)
                    numberOfViolations++
                    if violation.severity.isError {
                        numberOfSeriousViolations++
                    }
                }
            }
            let violationSuffix = (numberOfViolations != 1 ? "s" : "")
            let filesSuffix = (filesToLint.count != 1 ? "s." : ".")
            print(
                "Done linting!" +
                " Found \(numberOfViolations) violation\(violationSuffix)," +
                " \(numberOfSeriousViolations) serious" +
                " in \(filesToLint.count) file\(filesSuffix)"
            )
            if numberOfSeriousViolations <= 0 {
                return .Success()
            }
            return .Failure(CommandantError<()>.CommandError())
        }
        return .Failure(CommandantError<()>.UsageError(description: "No lintable files found at" +
            " path \(path)"))
    }

    private func filesToLintAtPath(path: String) -> [String] {
        let absolutePath = (path.absolutePathRepresentation() as NSString).stringByStandardizingPath
        var isDirectory: ObjCBool = false
        if fileManager.fileExistsAtPath(absolutePath, isDirectory: &isDirectory) {
            if isDirectory {
                return fileManager.allFilesRecursively(directory: absolutePath).filter {
                    $0.isSwiftFile()
                }
            } else if absolutePath.isSwiftFile() {
                return [absolutePath]
            }
        }
        return []
    }
}

struct LintOptions: OptionsType {
    let path: String
    let useSTDIN: Bool
    let configurationFile: String

    static func create(path: String)(useSTDIN: Bool)(configurationFile: String) -> LintOptions {
        return LintOptions(path: path, useSTDIN: useSTDIN, configurationFile: configurationFile)
    }

    static func evaluate(m: CommandMode) -> Result<LintOptions, CommandantError<()>> {
        return create
            <*> m <| Option(key: "path",
                defaultValue: "",
                usage: "the path to the file or directory to lint")
            <*> m <| Option(key: "use-stdin",
                defaultValue: false,
                usage: "lint standard input")
            <*> m <| Option(key: "config",
                defaultValue: ".swiftlint.yml",
                usage: "the path to SwiftLint's configuration file")
    }
}
